### PR TITLE
fix(opt): fix merge multi join

### PR DESCRIPTION
**Pull request by st1page**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

### CODE REVIEW

File: src/frontend/test_runner/tests/testdata/join.yaml [hunk 1]
```diff
@@ -47,36 +47,36 @@
...
-              BatchScan { table: t1, columns: [v1, v2] }
-            BatchExchange { order: [], dist: HashShard([0]) }
-              BatchScan { table: t2, columns: [v1, v2] }
-        BatchExchange { order: [], dist: HashShard([1]) }
-          BatchScan { table: t3, columns: [v1, v2] }
+      BatchHashJoin { type: Inner, predicate: $5 = $6, output_indices: [0, 1, 2, 3, 4, 5] }
+        BatchExchange { order: [], dist: HashShard([5]) }
+          BatchHashJoin { type: Inner, predicate: $1 = $4, output_indices: all }
+            BatchExchange { order: [], dist: HashShard([1]) }
+              BatchScan { table: t1, columns: [_row_id, v1, v2] }
```
Comment (skyzh) [inline src/frontend/test_runner/tests/testdata/join.yaml:54]: why there's row_id in batch_query?

Comment (chenzl25) [reply src/frontend/test_runner/tests/testdata/join.yaml:54]: BTW,  LogicalMultiJoin `as_reordered_left_deep_join` lost output_indices

Comment (st1page) [reply src/frontend/test_runner/tests/testdata/join.yaml:54]: fixing

**Review by chenzl25 — COMMENTED** [review]
LogicalMultiJoinBuilder `with_filter` method simply add the filter to conjunctions is inappropriate, because filter condition refers its input (maybe a project). I think MergeMultiJoinRule should transform correctly even without predicate pushdown as its prerequisite. So make sure when we turn down the predicate pushdown, this rule still work correctly.
```rust
    fn with_filter(plan: PlanRef) -> LogicalMultiJoinBuilder {
        let filter: &LogicalFilter = plan.as_logical_filter().unwrap();
        let mut builder = Self::new(filter.input());
        builder
            .conjunctions
            .extend(filter.predicate().conjunctions.clone());
        builder
    }
```

**Comment by st1page** [comment]
> LogicalMultiJoinBuilder `with_filter` method simply add the filter to conjunctions is inappropriate, because filter condition refers its input (maybe a project). I think MergeMultiJoinRule should transform correctly even without predicate pushdown as its prerequisite. So make sure when we turn down the predicate pushdown, this rule still work correctly.
> 
> ```rust
>     fn with_filter(plan: PlanRef) -> LogicalMultiJoinBuilder {
>         let filter: &LogicalFilter = plan.as_logical_filter().unwrap();
>         let mut builder = Self::new(filter.input());
>         builder
>             .conjunctions
>             .extend(filter.predicate().conjunctions.clone());
>         builder
>     }
> ```

Oops, maybe the `on_condition` in the join has the same issue too. I will fix it.

**Comment by chenzl25** [comment]
LGTM

**Review by chenzl25 — APPROVED** [review]
LGTM

**Pull request merged**